// bitcoin-ts/build/main/lib/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils/utils");
/**
 * @param bin the raw transaction from which to read the input
 * @param offset the offset at which the input begins
 */
exports.readTransactionInput = (bin, offset) => {
    const offsetAfterTxHash = offset + 32 /* sha256Hash */;
    const outpointTransactionHash = bin
        .slice(offset, offsetAfterTxHash)
        .reverse();
    const offsetAfterOutpointIndex = offsetAfterTxHash + 4 /* uint32 */;
    const outpointIndex = utils_1.binToNumberUint32LE(bin.subarray(offsetAfterTxHash, offsetAfterOutpointIndex));
    const { nextOffset: offsetAfterScriptLength, value: scriptLength } = utils_1.readBitcoinVarInt(bin, offsetAfterOutpointIndex);
    const offsetAfterScript = offsetAfterScriptLength + Number(scriptLength);
    const unlockingBytecode = bin.slice(offsetAfterScriptLength, offsetAfterScript);
    const nextOffset = offsetAfterScript + 4 /* uint32 */;
    const sequenceNumber = utils_1.binToNumberUint32LE(bin.subarray(offsetAfterScript, nextOffset));
    return {
        input: {
            outpointIndex,
            outpointTransactionHash,
            sequenceNumber,
            unlockingBytecode
        },
        nextOffset
    };
};
/**
 * Serialize a single input.
 * @param output the input to serialize
 */
exports.serializeInput = (input) => utils_1.flattenBinArray([
    input.outpointTransactionHash.slice().reverse(),
    utils_1.numberToBinUint32LE(input.outpointIndex),
    utils_1.bigIntToBitcoinVarInt(BigInt(input.unlockingBytecode.length)),
    input.unlockingBytecode,
    utils_1.numberToBinUint32LE(input.sequenceNumber)
]);
/**
 * Serialize a set of inputs for inclusion in a serialized transaction.
 *
 * Format: <BitcoinVarInt: input count> <serialized inputs>
 *
 * @param inputs the set of inputs to serialize
 */
exports.serializeInputs = (inputs) => utils_1.flattenBinArray([
    utils_1.bigIntToBitcoinVarInt(BigInt(inputs.length)),
    ...inputs.map(exports.serializeInput)
]);
/**
 * @param bin the raw transaction from which to read the output
 * @param offset the offset at which the output begins
 */
exports.readTransactionOutput = (bin, offset) => {
    const offsetAfterSatoshis = offset + 8 /* uint64 */;
    const satoshis = utils_1.binToBigIntUint64LE(bin.subarray(offset, offsetAfterSatoshis));
    const { nextOffset: offsetAfterScriptLength, value } = utils_1.readBitcoinVarInt(bin, offsetAfterSatoshis);
    const scriptLength = Number(value);
    const nextOffset = offsetAfterScriptLength + scriptLength;
    const lockingBytecode = scriptLength === 0
        ? new Uint8Array()
        : bin.slice(offsetAfterScriptLength, nextOffset);
    return {
        nextOffset,
        output: {
            lockingBytecode,
            satoshis
        }
    };
};
/**
 * Serialize a single output.
 * @param output the output to serialize
 */
exports.serializeOutput = (output) => utils_1.flattenBinArray([
    utils_1.bigIntToBinUint64LE(BigInt(output.satoshis)),
    utils_1.bigIntToBitcoinVarInt(BigInt(output.lockingBytecode.length)),
    output.lockingBytecode
]);
/**
 * Serialize a set of outputs for inclusion in a serialized transaction.
 *
 * Format: <BitcoinVarInt: output count> <serialized outputs>
 *
 * @param outputs the set of outputs to serialize
 */
exports.serializeOutputsForTransaction = (outputs) => utils_1.flattenBinArray([
    utils_1.bigIntToBitcoinVarInt(BigInt(outputs.length)),
    ...outputs.map(exports.serializeOutput)
]);
/**
 * TODO: document return type (note outpointTransactionHash is little-endian – most UIs display big-endian transaction hashes)
 *
 * Note: this method throws runtime errors when attempting to decode improperly
 * encoded transactions.
 *
 * @param bin the raw transaction to decode
 */
exports.deserializeTransaction = (bin) => {
    const version = utils_1.binToNumberUint32LE(bin.subarray(0, 4 /* uint32 */));
    const offsetAfterVersion = 4 /* uint32 */;
    const { nextOffset: offsetAfterInputCount, value: inputCount } = utils_1.readBitcoinVarInt(bin, offsetAfterVersion);
    // eslint-disable-next-line functional/no-let
    let cursor = offsetAfterInputCount;
    const inputs = [];
    // eslint-disable-next-line functional/no-let, functional/no-loop-statement, no-plusplus
    for (let i = 0; i < Number(inputCount); i++) {
        const { input, nextOffset } = exports.readTransactionInput(bin, cursor);
        // eslint-disable-next-line functional/no-expression-statement
        cursor = nextOffset;
        // eslint-disable-next-line functional/no-expression-statement, functional/immutable-data
        inputs.push(input);
    }
    const { nextOffset: offsetAfterOutputCount, value: outputCount } = utils_1.readBitcoinVarInt(bin, cursor);
    // eslint-disable-next-line functional/no-expression-statement
    cursor = offsetAfterOutputCount;
    const outputs = [];
    // eslint-disable-next-line functional/no-let, functional/no-loop-statement, no-plusplus
    for (let i = 0; i < Number(outputCount); i++) {
        const { output, nextOffset } = exports.readTransactionOutput(bin, cursor);
        // eslint-disable-next-line functional/no-expression-statement
        cursor = nextOffset;
        // eslint-disable-next-line functional/no-expression-statement, functional/immutable-data
        outputs.push(output);
    }
    const locktime = utils_1.binToNumberUint32LE(bin.subarray(cursor, cursor + 4 /* uint32 */));
    return {
        inputs,
        locktime,
        outputs,
        version
    };
};
/**
 * TODO: doc
 */
exports.serializeTransaction = (tx) => utils_1.flattenBinArray([
    utils_1.numberToBinUint32LE(tx.version),
    exports.serializeInputs(tx.inputs),
    exports.serializeOutputsForTransaction(tx.outputs),
    utils_1.numberToBinUint32LE(tx.locktime)
]);
/**
 * Derive a standard identifier from a serialized data structure.
 *
 * @remarks
 * By convention, Bitcoin transaction and block identifiers are derived by
 * double-sha256 hashing their serialized form, and reversing the byte order.
 * (The result of sha256 is defined by its specification as big-endian, and
 * bitcoin displays hashes in little-endian format.)
 *
 * @returns an identifier in little-endian byte order
 *
 * @param data the serialized raw data being identified
 * @param sha256 an implementation of sha256
 */
exports.getBitcoinIdentifier = (data, sha256) => sha256.hash(sha256.hash(data)).reverse();
/**
 * Derive a standard transaction identifier from a serialized transaction.
 *
 * @returns a Transaction ID in little-endian byte order
 *
 * @param transaction the serialized transaction
 * @param sha256 an implementation of sha256
 */
exports.getBitcoinTransactionId = (transaction, sha256) => utils_1.binToHex(exports.getBitcoinIdentifier(transaction, sha256));
/**
 * Get the hash of all outpoints in a series of inputs. (For use in
 * `hashTransactionOutpoints`.)
 *
 * @param inputs the series of inputs from which to extract the outpoints
 * @param sha256 an implementation of sha256
 */
exports.serializeOutpoints = (inputs) => utils_1.flattenBinArray(inputs.map(i => utils_1.flattenBinArray([
    i.outpointTransactionHash.slice().reverse(),
    utils_1.numberToBinUint32LE(i.outpointIndex)
])));
/**
 * Get the signing serialization for a series of outputs.
 * @param outputs the series of outputs to serialize
 */
exports.serializeOutputsForSigning = (outputs) => utils_1.flattenBinArray(outputs.map(exports.serializeOutput));
/**
 * Serialize a series of input sequence numbers.
 *
 * @param inputs the series of inputs from which to extract the sequence numbers
 */
exports.serializeSequenceNumbers = (inputs) => utils_1.flattenBinArray(inputs.map(i => utils_1.numberToBinUint32LE(i.sequenceNumber)));
