// bitcoin-ts/build/module/lib/auth/instruction-sets/common/crypto.js
import { ConsensusBCH } from '../bch/bch-types';
import { serializeAuthenticationInstructions } from '../instruction-sets-utils';
import { combineOperations, pushToStack, useOneScriptNumber, useOneStackItem, useTwoStackItems } from './combinators';
import { booleanToScriptNumber, ConsensusCommon } from './common';
import { decodeBitcoinSignature, isValidPublicKeyEncoding, isValidSignatureEncodingBCHTransaction } from './encoding';
import { applyError, AuthenticationErrorCommon } from './errors';
import { opVerify } from './flow-control';
import { OpcodesCommon } from './opcodes';
import { generateSigningSerializationBCH } from './signing-serialization';
export const opRipemd160 = (ripemd160) => (state) => useOneStackItem(state, (nextState, value) => pushToStack(nextState, ripemd160.hash(value)));
export const opSha1 = (sha1) => (state) => useOneStackItem(state, (nextState, value) => pushToStack(nextState, sha1.hash(value)));
export const opSha256 = (sha256) => (state) => useOneStackItem(state, (nextState, value) => pushToStack(nextState, sha256.hash(value)));
export const opHash160 = (sha256, ripemd160) => (state) => useOneStackItem(state, (nextState, value) => pushToStack(nextState, ripemd160.hash(sha256.hash(value))));
export const opHash256 = (sha256) => (state) => useOneStackItem(state, (nextState, value) => pushToStack(nextState, sha256.hash(sha256.hash(value))));
export const opCodeSeparator = () => (state) => {
    // eslint-disable-next-line functional/no-expression-statement, functional/immutable-data
    state.lastCodeSeparator = state.ip;
    return state;
};
export const opCheckSig = (sha256, secp256k1, flags) => (s) => 
// eslint-disable-next-line complexity
useTwoStackItems(s, (state, bitcoinEncodedSignature, publicKey) => {
    if (!isValidPublicKeyEncoding(publicKey)) {
        return applyError(AuthenticationErrorCommon.invalidPublicKeyEncoding, state);
    }
    if (!isValidSignatureEncodingBCHTransaction(bitcoinEncodedSignature)) {
        return applyError(AuthenticationErrorCommon.invalidSignatureEncoding, state);
    }
    const coveredBytecode = serializeAuthenticationInstructions(state.instructions).subarray(state.lastCodeSeparator + 1);
    const { signingSerializationType, signature } = decodeBitcoinSignature(bitcoinEncodedSignature);
    const serialization = generateSigningSerializationBCH(sha256, state.version, state.transactionOutpoints, state.transactionSequenceNumbers, state.outpointTransactionHash, state.outpointIndex, coveredBytecode, state.outputValue, state.sequenceNumber, state.correspondingOutput, state.transactionOutputs, state.locktime, signingSerializationType);
    const digest = sha256.hash(sha256.hash(serialization));
    const useSchnorr = signature.length === ConsensusBCH.schnorrSignatureLength;
    const success = useSchnorr
        ? secp256k1.verifySignatureSchnorr(signature, publicKey, digest)
        : secp256k1.verifySignatureDERLowS(signature, publicKey, digest);
    return !success &&
        flags.requireNullSignatureFailures &&
        signature.length !== 0
        ? applyError(AuthenticationErrorCommon.nonNullSignatureFailure, state)
        : pushToStack(state, booleanToScriptNumber(success));
});
export const opCheckMultiSig = (sha256, secp256k1, flags) => (s) => useOneScriptNumber(s, (state, publicKeysValue) => {
    const potentialPublicKeys = Number(publicKeysValue);
    if (potentialPublicKeys < 0) {
        return applyError(AuthenticationErrorCommon.invalidNaturalNumber, state);
    }
    if (potentialPublicKeys > 20 /* maximumPublicKeys */) {
        return applyError(AuthenticationErrorCommon.exceedsMaximumMultisigPublicKeyCount, state);
    }
    const publicKeys = 
    // eslint-disable-next-line functional/immutable-data
    potentialPublicKeys > 0 ? state.stack.splice(-potentialPublicKeys) : [];
    // eslint-disable-next-line functional/no-expression-statement, functional/immutable-data
    state.operationCount += potentialPublicKeys;
    return state.operationCount > ConsensusCommon.maximumOperationCount
        ? applyError(AuthenticationErrorCommon.exceededMaximumOperationCount, state)
        : useOneScriptNumber(state, (nextState, approvingKeys) => {
            const requiredApprovingPublicKeys = Number(approvingKeys);
            if (requiredApprovingPublicKeys < 0) {
                return applyError(AuthenticationErrorCommon.invalidNaturalNumber, nextState);
            }
            if (requiredApprovingPublicKeys > potentialPublicKeys) {
                return applyError(AuthenticationErrorCommon.insufficientPublicKeys, nextState);
            }
            const signatures = requiredApprovingPublicKeys > 0
                ? // eslint-disable-next-line functional/immutable-data
                    nextState.stack.splice(-requiredApprovingPublicKeys)
                : [];
            return useOneStackItem(nextState, 
            // eslint-disable-next-line complexity
            (finalState, protocolBugValue) => {
                if (flags.requireBugValueZero &&
                    protocolBugValue.length !== 0) {
                    return applyError(AuthenticationErrorCommon.invalidProtocolBugValue, finalState);
                }
                const coveredBytecode = serializeAuthenticationInstructions(finalState.instructions).subarray(finalState.lastCodeSeparator + 1);
                let approvingPublicKeys = 0; // eslint-disable-line functional/no-let
                let remainingSignatures = signatures.length; // eslint-disable-line functional/no-let
                let remainingPublicKeys = publicKeys.length; // eslint-disable-line functional/no-let
                // eslint-disable-next-line functional/no-loop-statement
                while (remainingSignatures > 0 &&
                    remainingPublicKeys > 0 &&
                    approvingPublicKeys + remainingPublicKeys >=
                        remainingSignatures &&
                    approvingPublicKeys !== requiredApprovingPublicKeys) {
                    const publicKey = publicKeys[remainingPublicKeys - 1];
                    const bitcoinEncodedSignature = signatures[remainingSignatures - 1];
                    if (!isValidPublicKeyEncoding(publicKey)) {
                        return applyError(AuthenticationErrorCommon.invalidPublicKeyEncoding, finalState);
                    }
                    if (!isValidSignatureEncodingBCHTransaction(bitcoinEncodedSignature)) {
                        return applyError(AuthenticationErrorCommon.invalidSignatureEncoding, finalState);
                    }
                    const { signingSerializationType, signature } = decodeBitcoinSignature(bitcoinEncodedSignature);
                    const serialization = generateSigningSerializationBCH(sha256, finalState.version, finalState.transactionOutpoints, finalState.transactionSequenceNumbers, finalState.outpointTransactionHash, finalState.outpointIndex, coveredBytecode, finalState.outputValue, finalState.sequenceNumber, finalState.correspondingOutput, finalState.transactionOutputs, finalState.locktime, signingSerializationType);
                    const digest = sha256.hash(sha256.hash(serialization));
                    if (signature.length === ConsensusBCH.schnorrSignatureLength) {
                        return applyError(AuthenticationErrorCommon.schnorrSizedSignatureInCheckMultiSig, finalState);
                    }
                    const signed = secp256k1.verifySignatureDERLowS(signature, publicKey, digest);
                    // eslint-disable-next-line functional/no-conditional-statement
                    if (signed) {
                        approvingPublicKeys += 1; // eslint-disable-line functional/no-expression-statement
                        remainingSignatures -= 1; // eslint-disable-line functional/no-expression-statement
                    }
                    remainingPublicKeys -= 1; // eslint-disable-line functional/no-expression-statement
                }
                const success = approvingPublicKeys === requiredApprovingPublicKeys;
                if (!success &&
                    flags.requireNullSignatureFailures &&
                    !signatures.every(signature => signature.length === 0)) {
                    return applyError(AuthenticationErrorCommon.nonNullSignatureFailure, finalState);
                }
                return pushToStack(finalState, booleanToScriptNumber(success));
            });
        }, flags.requireMinimalEncoding);
}, flags.requireMinimalEncoding);
export const opCheckSigVerify = (sha256, secp256k1, flags) => combineOperations(opCheckSig(sha256, secp256k1, flags), opVerify());
export const opCheckMultiSigVerify = (sha256, secp256k1, flags) => combineOperations(opCheckMultiSig(sha256, secp256k1, flags), opVerify());
export const cryptoOperations = (sha1, sha256, ripemd160, secp256k1, flags) => ({
    [OpcodesCommon.OP_RIPEMD160]: opRipemd160(ripemd160),
    [OpcodesCommon.OP_SHA1]: opSha1(sha1),
    [OpcodesCommon.OP_SHA256]: opSha256(sha256),
    [OpcodesCommon.OP_HASH160]: opHash160(sha256, ripemd160),
    [OpcodesCommon.OP_HASH256]: opHash256(sha256),
    [OpcodesCommon.OP_CODESEPARATOR]: opCodeSeparator(),
    [OpcodesCommon.OP_CHECKSIG]: opCheckSig(sha256, secp256k1, flags),
    [OpcodesCommon.OP_CHECKSIGVERIFY]: opCheckSigVerify(sha256, secp256k1, flags),
    [OpcodesCommon.OP_CHECKMULTISIG]: opCheckMultiSig(sha256, secp256k1, flags),
    [OpcodesCommon.OP_CHECKMULTISIGVERIFY]: opCheckMultiSigVerify(sha256, secp256k1, flags)
});
