// bitcoin-ts/build/main/lib/auth/instruction-sets/bch/bch-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const transaction_1 = require("../../../transaction");
const utils_1 = require("../../../utils/utils");
var ConsensusBCH;
(function (ConsensusBCH) {
    ConsensusBCH[ConsensusBCH["schnorrSignatureLength"] = 64] = "schnorrSignatureLength";
})(ConsensusBCH = exports.ConsensusBCH || (exports.ConsensusBCH = {}));
exports.createTestAuthenticationProgramBCH = (unlockingBytecode, lockingBytecode, sha256, satoshis = BigInt(0)) => {
    const testFundingTransaction = {
        inputs: [
            {
                outpointIndex: 0xffffffff,
                outpointTransactionHash: utils_1.hexToBin('0000000000000000000000000000000000000000000000000000000000000000'),
                sequenceNumber: 0xffffffff,
                unlockingBytecode: Uint8Array.of(0, 0)
            }
        ],
        locktime: 0,
        outputs: [{ lockingBytecode, satoshis }],
        version: 1
    };
    const testSpendingTransaction = {
        inputs: [
            {
                outpointIndex: 0,
                outpointTransactionHash: utils_1.hexToBin(utils_1.swapEndianness(transaction_1.getBitcoinTransactionId(transaction_1.serializeTransaction(testFundingTransaction), sha256))),
                sequenceNumber: 0xffffffff,
                unlockingBytecode
            }
        ],
        locktime: 0,
        outputs: [{ lockingBytecode: Uint8Array.of(), satoshis }],
        version: 1
    };
    return {
        inputIndex: 0,
        sourceOutput: testFundingTransaction.outputs[0],
        spendingTransaction: testSpendingTransaction
    };
};
