// bitcoin-ts/build/main/lib/crypto/ripemd160.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bin_1 = require("../bin/bin");
const utils_1 = require("../utils/utils");
/**
 * The most performant way to instantiate ripemd160 functionality. To avoid
 * using Node.js or DOM-specific APIs, you can use `instantiateRipemd160`.
 *
 * @param webassemblyBytes A buffer containing the ripemd160 binary.
 */
exports.instantiateRipemd160Bytes = async (webassemblyBytes) => {
    const wasm = await bin_1.instantiateRustWasm(webassemblyBytes, './ripemd160', 'ripemd160', 'ripemd160_init', 'ripemd160_update', 'ripemd160_final');
    return {
        final: wasm.final,
        hash: wasm.hash,
        init: wasm.init,
        update: wasm.update
    };
};
exports.getEmbeddedRipemd160Binary = () => utils_1.base64ToBin(bin_1.ripemd160Base64Bytes).buffer;
/**
 * An ultimately-portable (but slower) version of `instantiateRipemd160Bytes`
 * which does not require the consumer to provide the ripemd160 binary buffer.
 */
exports.instantiateRipemd160 = async () => exports.instantiateRipemd160Bytes(exports.getEmbeddedRipemd160Binary());
