// bitcoin-ts/build/module/lib/auth/templates/language/reduce.js
import { flattenBinArray } from '../../../utils/utils';
import { OpcodesCommon } from '../../instruction-sets/common/opcodes';
import { encodeDataPush } from '../../instruction-sets/instruction-sets';
import { authenticationInstructionsAreNotMalformed, disassembleBytecode, parseBytecode } from '../../instruction-sets/instruction-sets-utils';
const pluckStartPosition = (range) => ({
    startColumn: range.startColumn,
    startLineNumber: range.startLineNumber
});
const pluckEndPosition = (range) => ({
    endColumn: range.endColumn,
    endLineNumber: range.endLineNumber
});
const mergeRanges = (ranges) => {
    const unsortedMerged = ranges.reduce(
    // eslint-disable-next-line complexity
    (merged, range) => ({
        ...(range.startLineNumber < merged.startLineNumber
            ? pluckStartPosition(range)
            : range.startLineNumber === merged.startLineNumber &&
                range.startColumn < merged.startColumn
                ? pluckStartPosition(range)
                : pluckStartPosition(merged)),
        ...(range.endLineNumber > merged.endLineNumber
            ? pluckEndPosition(range)
            : range.endLineNumber === merged.endLineNumber &&
                range.endColumn > merged.endColumn
                ? pluckEndPosition(range)
                : pluckEndPosition(merged))
    }), ranges[0]);
    return {
        ...pluckStartPosition(unsortedMerged),
        ...pluckEndPosition(unsortedMerged)
    };
};
const emptyReductionTraceNode = (range) => ({
    bytecode: Uint8Array.of(),
    range
});
/**
 * Aggregate instructions to build groups of non-malformed instructions.
 *
 * --- TODO: delete old stuff below? ––-
 *
 * So users can write constructions like `OP_PUSHBYTES_2 0x0102` which will
 * evaluate successfully, but instructions which must read over a new-line to
 * make sense (e.g. `OP_PUSHBYTES_2\n0x0102`) should error. This allows for a
 * nice omniscient-debugging experience.
 *
 * **This makes new lines important in evaluations.** However, things
 * can only "break" when new lines are inserted, not when they are removed (e.g.
 * if a script is "minified" to a single line for deployment.)
 *
 * **Implementation note**
 * This method aggregates arrays of instructions by line... a little like
 * Automatic Semicolon Insertion in ECMAScript. 👀 In fact, that's a good
 * sign that we're missing a useful language construct here. Maybe instead
 * of this algorithm, we need a new type of wrapper in the language to indicate
 * that bytecode segments are intended to go together.
 *
 * Interestingly, we already use "Containers" in both pushes and evaluations, so
 * this might be quite easy. E.g. wrapping with `()` or `{}`. However, we also
 * want disassembled instructions to be valid input in BTL, so some form of this
 * line-based logic will still be required unless we also change script
 * disassembly form. E.g. instead of `OP_PUSHBYTES_2 0x0102`, something like
 * `(OP_PUSHBYTES_2 0x0102)` or `<0x0102>`. This is something to consider in
 * future versions.
 */
// eslint-disable-next-line complexity
const aggregatedParseReductionTraceNodes = (nodes) => {
    const aggregations = [];
    // eslint-disable-next-line functional/no-let
    let ip = 0;
    // eslint-disable-next-line functional/no-let, init-declarations
    let incomplete;
    // eslint-disable-next-line functional/no-loop-statement
    for (const node of nodes) {
        const bytecode = incomplete === undefined
            ? node.bytecode
            : flattenBinArray([incomplete.bytecode, node.bytecode]);
        const range = incomplete === undefined
            ? node.range
            : mergeRanges([incomplete.range, node.range]);
        // eslint-disable-next-line functional/no-expression-statement
        incomplete = undefined;
        const parsed = parseBytecode(bytecode);
        // eslint-disable-next-line functional/no-conditional-statement
        if (parsed.length === 0) {
            // eslint-disable-next-line functional/no-expression-statement, functional/immutable-data
            aggregations.push({
                instructions: [],
                lastIp: ip,
                range
            });
            // eslint-disable-next-line functional/no-conditional-statement
        }
        else if (authenticationInstructionsAreNotMalformed(parsed)) {
            // eslint-disable-next-line functional/no-expression-statement
            ip += parsed.length;
            // eslint-disable-next-line functional/no-expression-statement, functional/immutable-data
            aggregations.push({
                instructions: parsed,
                lastIp: ip,
                range
            });
            // eslint-disable-next-line functional/no-conditional-statement
        }
        else {
            // eslint-disable-next-line functional/no-expression-statement
            incomplete = { bytecode, range };
        }
    }
    return {
        aggregations,
        success: true,
        ...(incomplete === undefined
            ? undefined
            : {
                remainingBytecode: incomplete.bytecode,
                remainingRange: incomplete.range,
                success: false
            })
    };
};
/**
 * Evaluate an array of `InstructionAggregation`s with the provided
 * `AuthenticationVirtualMachine`, matching the results back to their source
 * ranges.
 */
export const evaluateInstructionAggregations = (aggregations, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
vm, getState) => {
    const nonEmptyAggregations = aggregations.filter(aggregation => aggregation.instructions.length > 0);
    const evaluationPlan = nonEmptyAggregations.reduce((plan, aggregation) => {
        const instructions = [...plan.instructions, ...aggregation.instructions];
        return {
            breakpoints: [
                ...plan.breakpoints,
                { ip: aggregation.lastIp, range: aggregation.range }
            ],
            instructions
        };
    }, { breakpoints: [], instructions: [] });
    const trace = vm.stateDebug(getState(evaluationPlan.instructions));
    const samples = evaluationPlan.breakpoints.map(breakpoint => ({
        range: breakpoint.range,
        state: trace[breakpoint.ip - 1]
    }));
    const firstInvalidSample = samples.findIndex(sample => sample.state === undefined);
    const errorSample = samples[firstInvalidSample - 1] ??
        samples[firstInvalidSample];
    return errorSample === undefined
        ? {
            samples: samples,
            success: true
        }
        : {
            errors: [
                {
                    error: errorSample.state === undefined
                        ? `Failed to reduce evaluation: vm.debug produced no valid program states.`
                        : `Failed to reduce evaluation: ${errorSample.state.error ??
                            'unknown error'}`,
                    range: errorSample.range
                }
            ],
            samples,
            success: false
        };
};
/**
 * Incrementally evaluate an array of `ScriptReductionTraceNode`s, returning a
 * trace of the evaluation and the resulting top stack item (`evaluationResult`)
 * if successful.
 *
 * @param nodes an array of reduced nodes
 * @param vm the `AuthenticationVirtualMachine` to use in the evaluation
 * @param getState a method which should generate a new ProgramState given an
 * array of `instructions`
 */
// eslint-disable-next-line complexity
export const sampledEvaluateReductionTraceNodes = (nodes, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
vm, getState) => {
    const parsed = aggregatedParseReductionTraceNodes(nodes);
    const evaluated = evaluateInstructionAggregations(parsed.aggregations, vm, getState);
    if (parsed.success && evaluated.success) {
        const samples = evaluated.samples.length > 0
            ? evaluated.samples
            : [{ range: parsed.aggregations[0].range, state: getState([]) }];
        const lastSample = samples[samples.length - 1];
        const lastStackItem = lastSample.state.stack[lastSample.state.stack.length - 1];
        const evaluationResult = lastStackItem === undefined ? Uint8Array.of() : lastStackItem.slice();
        return {
            bytecode: evaluationResult,
            samples,
            success: true
        };
    }
    return {
        bytecode: Uint8Array.of(),
        errors: [
            ...(parsed.success
                ? []
                : [
                    {
                        error: `A sample is malformed and cannot be evaluated: ${disassembleBytecode(OpcodesCommon, parsed.remainingBytecode)}`,
                        range: parsed.remainingRange
                    }
                ]),
            ...(evaluated.success ? [] : evaluated.errors)
        ],
        samples: evaluated.samples,
        success: false
    };
};
/**
 * This method will throw an error if provided a `compiledScript` with
 * compilation errors. To check for compilation errors, use `getCompileErrors`.
 * @param compiledScript the `CompiledScript` to reduce
 * @param vm the `AuthenticationVirtualMachine` to use for evaluations
 * @param createState a method which returns the base `ProgramState` used when initializing evaluations
 */
export const reduceScript = (compiledScript, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
vm, createState) => {
    const source = compiledScript.map(segment => {
        switch (segment.type) {
            case 'bytecode':
                return { bytecode: segment.value, range: segment.range };
            case 'push': {
                if (segment.value.length === 0) {
                    return emptyReductionTraceNode(segment.range);
                }
                const push = reduceScript(segment.value, vm, createState);
                const bytecode = encodeDataPush(push.bytecode);
                return {
                    bytecode,
                    ...(push.errors === undefined ? undefined : { errors: push.errors }),
                    range: segment.range,
                    source: [push]
                };
            }
            case 'evaluation': {
                if (segment.value.length === 0) {
                    return emptyReductionTraceNode(segment.range);
                }
                if (typeof vm === 'undefined' || typeof createState === 'undefined') {
                    return {
                        errors: [
                            {
                                error: 'Both a VM and a createState method are required to reduce evaluations.',
                                range: segment.range
                            }
                        ],
                        ...emptyReductionTraceNode(segment.range)
                    };
                }
                const reductionTrace = reduceScript(segment.value, vm, createState);
                const evaluated = sampledEvaluateReductionTraceNodes(reductionTrace.source, vm, createState);
                const errors = [
                    ...(reductionTrace.errors === undefined ? [] : reductionTrace.errors),
                    ...(evaluated.success ? [] : evaluated.errors)
                ];
                return {
                    ...(errors.length > 0
                        ? {
                            errors,
                            ...emptyReductionTraceNode(segment.range)
                        }
                        : {
                            bytecode: evaluated.bytecode,
                            range: segment.range
                        }),
                    samples: evaluated.samples,
                    source: [reductionTrace]
                };
            }
            case 'comment':
                return emptyReductionTraceNode(segment.range);
            case 'error':
                return {
                    errors: [
                        {
                            error: `Tried to reduce a BTL script with resolution errors: ${segment.value}`,
                            range: segment.range
                        }
                    ],
                    ...emptyReductionTraceNode(segment.range)
                };
            default:
                return new Error(
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                `"${segment.type}" is not a known segment type.`);
        }
    });
    const reduction = source.reduce((all, segment) => ({
        bytecode: [...all.bytecode, segment.bytecode],
        ranges: [...all.ranges, segment.range],
        ...(all.errors !== undefined || segment.errors !== undefined
            ? {
                errors: [
                    ...(all.errors === undefined ? [] : all.errors),
                    ...(segment.errors === undefined ? [] : segment.errors)
                ]
            }
            : undefined)
    }), { bytecode: [], ranges: [] });
    return {
        ...(reduction.errors === undefined
            ? undefined
            : { errors: reduction.errors }),
        bytecode: flattenBinArray(reduction.bytecode),
        range: mergeRanges(reduction.ranges),
        source
    };
};
